File: Monotonic/LongestSubarray.java

/*
Given an array of integers nums and an integer limit,
return the size of the longest subarray
such that the absolute difference between any two elements of this subarray
is less than or equal to limit.
*/

import java.util.ArrayDeque;

public class LongestSubarray {
    public static int longestSubarray(int[] nums, int limit) {
        ArrayDeque<Integer> monoIncreasing = new ArrayDeque<>();
        ArrayDeque<Integer> monoDecreasing = new ArrayDeque<>();
        int left = 0;
        int maxSubarray = 0;

        for (int right = 0; right < nums.length; right++) {
            // Maintain monotonic queues
            while (!monoIncreasing.isEmpty() && nums[right] < nums[monoIncreasing.getLast()]) {
                monoIncreasing.removeLast();
            }
            while (!monoDecreasing.isEmpty() && nums[right] > nums[monoDecreasing.getLast()]) {
                monoDecreasing.removeLast();
            }

            monoIncreasing.addLast(right);
            monoDecreasing.addLast(right);

            // Maintain valid window
            while (nums[monoDecreasing.getFirst()] - nums[monoIncreasing.getFirst()] > limit) {
                left++;
                if (monoDecreasing.getFirst() < left) {
                    monoDecreasing.removeFirst();
                }
                if (monoIncreasing.getFirst() < left) {
                    monoIncreasing.removeFirst();
                }
            }

            maxSubarray = Math.max(maxSubarray, right - left + 1);
            
        }


        return maxSubarray;
    }

    public static void main(String[] args) {
        int[] nums = {8,2,4,7};
        int limit = 4;


        System.out.println(longestSubarray(nums, limit));
    }
    
}

// Time complexity: O(n)
// Space complexity: O(n) in the worst case when all elements are the same and both queues grow to the size of the input array.

File: Monotonic/longest_subarray.py
from collections import deque
from typing import List


class LongestSubarray:
    def longest_subarray(self, nums: List[int], limit: int) -> int:
        monoDecreasing = deque()
        monoIncreasing = deque()
        max_window = 0

        left = 0
        for right in range(len(nums)):
            num = nums[right]

            while monoDecreasing and num > nums[monoDecreasing[-1]]:
                monoDecreasing.pop()
            monoDecreasing.append(right)

            while monoIncreasing and num < nums[monoIncreasing[-1]]:
                monoIncreasing.pop()
            monoIncreasing.append(right)

            # Maintain valid window
            while nums[monoDecreasing[0]] - nums[monoIncreasing[0]] > limit:
                left += 1
                if monoDecreasing[0] < left:
                    monoDecreasing.popleft()
                if monoIncreasing[0] < left:
                    monoIncreasing.popleft()

            max_window = max(right - left + 1, max_window)

        return max_window


# Time Complexity: O(n) where n is the number of elements in the input array nums, as each element is processed at most twice (once when added to the deques and once when removed).
# Space Complexity: O(n) in the worst case, when the deques contain all the elements of the current window.
